### ISSUE TITLE
[Feature] support `gather` instead of `all_gather` when gathering the logits

### ISSUE THREAD

**Issue by chunyuan-w** [sgl-project/sglang#3365]
### Checklist

- [x] 1. If the issue you raised is not a feature but a question, please raise a discussion at https://github.com/sgl-project/sglang/discussions/new/choose Otherwise, it will be closed.
- [x] 2. Please use English, otherwise it will be closed.

### Motivation

We noticed that in the `_get_logits` function of vllm, `gather` instead of `all_gather` will be used under certain conditions (the main condition is that for non-tpu devices):
Code link:

- [logits = tensor_model_parallel_gather(logits)](https://github.com/vllm-project/vllm/blob/6e1fc61f0fb90c37f0d4a1a8f76235a6e4e1103c/vllm/model_executor/layers/logits_processor.py#L101C22-L101C50)

- [condition of whether using `all_gather` or `gather`](https://github.com/vllm-project/vllm/blob/6e1fc61f0fb90c37f0d4a1a8f76235a6e4e1103c/vllm/model_executor/layers/logits_processor.py#L53-L57)

The change from using `all_gather` to `gather` is initially added in this PR for your reference: https://github.com/vllm-project/vllm/pull/2221.

While in SGLang, we see currently `all_gather` is always used:
https://github.com/sgl-project/sglang/blob/e868d0b60eb2d435c5599165f787bca06bdc9c3d/python/sglang/srt/layers/logits_processor.py#L246

Does SGLang have the plan to add `gather` instead of only `all_gather` when gathering the logits? Per the practice in vllm, using `gather` seems to have better performance than `all_gather` on devices which have `gather` support.

**Comment by zhaochenyang20** [comment]
Good idea. We should have someone working on this.

**Comment by voidxb** [comment]
I wanna own this issue @zhaochenyang20

**Comment by zhaochenyang20** [comment]
@voidxb Sure. Could have a try.

**Comment by voidxb** [comment]
https://github.com/vllm-project/vllm/pull/12071/files in this vllm pr, I see the use_all_gather var but firstly, I see `parallel_config = get_current_vllm_config().parallel_config self.use_all_gather = current_platform.is_tpu() \ or envs.VLLM_USE_V1 \ or parallel_config.distributed_executor_backend == "external_launcher" # noqa` so it seems that in new v1 arch, vllm uses all_gather firstly. 

then from [vllm/distributed/device_communicators/base_device_communicator.py](https://github.com/vllm-project/vllm/blob/9bebc9512f9340e94579b9bd69cfdc452c4d5bb0/vllm/distributed/device_communicators/base_device_communicator.py) and [vllm/model_executor/layers/logits_processor.py](https://github.com/vllm-project/vllm/blob/v0.7.3/vllm/model_executor/layers/logits_processor.py#L101), I found logits could be return None but in [model_runner](https://github.com/vllm-project/vllm/blob/v0.7.3/vllm/worker/model_runner.py#L1772) , logits just is delivered to sampler and sampler.forward does not allow logits could be None(assert). [@chunyuan-w](https://github.com/chunyuan-w) so do you really use this feature in production? in vllm and it really works well? 

I do a version to implement it, but its output seems bad(without token broadcast, just like vllm), and in my opinion, I think it might should change more code to let workers(tp_rank >0) could handle None logits. 

and as I could see, if we don't use all_gather in get_logits, tprank > 0 worker should get token from rank 0 worker as well, so we should broadcast token from rank0 to other ranks.In all_gather, other rank get token from logits(after sampling), but in gather method, we have to  add a new path to get token.
so Is It really an opt? 
the fix (use gather instead of all_gather ) in this pr is not the core fetaure and was merged in Jan this year, so I have some doubts.
Or could a logits gather & a token broadcast obviously decrease communication cost(compare to a logits all gather)?

thanks

**Comment by zhaochenyang20** [comment]
cc @FrankLeeeee

**Comment by chunyuan-w** [comment]
Hi @voidxb thanks for looking into this.

Yes we do use `gather` and it works well.
[VLLM_USE_V1](https://github.com/vllm-project/vllm/blob/v0.7.3/vllm/envs.py#L76) is still turned off by default so we're not running into V1 currently.
When using gather, only the driver worker (rank 0 of tensor parallel group) will perform sampling, the other workers directly `return []` [here](https://github.com/vllm-project/vllm/blob/v0.7.3/vllm/worker/model_runner.py#L1775-L1776) and won't call the `self.model.sample`.

**Comment by voidxb** [comment]
ok, I have see the difference
in vllm, before forward,  it broadcasts model input to other ranks
and only rank0(driver worker) do sampling
so in that case, all_gather logits is unnecessary(token broadcast + logits all gather -> token broadcast  + logits gather)
but in sglang, we have no token broadcast step, so all gather is necessary now
I will try to add token broadcast step

**Comment by zhaochenyang20** [comment]
@voidxb Thankss so muchc!

**Comment by ramgandikota** [comment]
Hello @voidxb are you still working on this? If not, would love to pick it up and create a PR.
